### ISSUE TITLE
Add drwrap mode that inverts control for use inside a drbbdup case

### ISSUE THREAD

**Issue by derekbruening** [DynamoRIO/dynamorio#5356]
Split from #3995

Pasting from there:

I guess this raises a larger issue: we want the drwrap clean call to only
be inserted for one of our drbbdup cases. But with no control over drwrap,
we can't arrange that, and drwrap is going to go insert its clean call at
the start before the drbbdup case dispatch I would guess.

How to solve?

If drbbdup were integrated inside drmgr: would non-bbdup-aware users of
drmgr like drwrap be invoked for the default case insertion and not the
other cases? So the user can either have wrapping for just one case (has
to be default)? Provide an option where user can pick one case, or all
cases?

Or, drwrap has to turn its control model around and have the user call
"do_drwrap_instru" from its insertion event, with drwrap not registering
instru events (but still registering modload, etc.)?

Leaning toward the latter: add a drwrap global flag.

However, we have several issues implementing this:

+ There is no app2app event inside drbbdup!
  => For now, don't support drwrap_replace.
+ The "for_trace" and "translating" params are not passed
  => For now, document that drwrap doesn't need them
+ The dr_emit_flags_t return value is not propagated
  => For now, document that when wrapping and not replacing it always
  returns DR_EMIT_DEFAULT

We would have to augment drbbdup to support drwrap_replace which needs app2app
and under some options needs to return something other than DR_EMIT_DEFAULT.

**Comment by derekbruening** [comment]
> * The dr_emit_flags_t return value is not propagated
>   => For now, document that when wrapping and not replacing it always
>   returns DR_EMIT_DEFAULT

I had this wrong: drwrap_wrap does have a case where it does not return DR_EMIT_DEFAULT.
For DRWRAP_REPLACE_RETADDR it returns DR_EMIT_MUST_END_TRACE.

@johnfxgalea -- do you see any issue with adding a dr_emit_flags_t return value to the 
drbbdup_options_t.instrument_instr callback (and having drbbdup bitwise-or them together)?
To not break compatibility: it would have to be a separate instrument_instr_ex or something.

I'm also hitting another problem with DRWRAP_REPLACE_RETADDR: it has a
replace_retaddr_sentinel which after the app2app phase has no app instrs and just a meta nop.
This breaks drbbdup's assumptions:
```
Program received signal SIGSEGV, Segmentation fault.
0x00005555557b4010 in instr_get_translation (instr=0x0) at /home/bruening/dr/git/src/core/ir/instr_shared.c:1147
1147	    return instr->translation;
(gdb) bt
#0  0x00005555557b4010 in instr_get_translation (instr=0x0) at /home/bruening/dr/git/src/core/ir/instr_shared.c:1147
#1  0x00005555557b276f in instr_get_app_pc (instr=0x0) at /home/bruening/dr/git/src/core/ir/instr_shared.c:482
#2  0x00005555559006ea in drbbdup_duplicate_phase (drcontext=0x7ffdf7afd080, tag=0x5555558f0051 <replace_retaddr_sentinel>, bb=0x7ffdf7b23138, 
    for_trace=0 '\000', translating=0 '\000') at /home/bruening/dr/git/src/ext/drbbdup/drbbdup.c:424
#3  0x000055555590c27a in drmgr_bb_event_do_instrum_phases (drcontext=0x7ffdf7afd080, tag=0x5555558f0051 <replace_retaddr_sentinel>, bb=0x7ffdf7b23138, 
    for_trace=0 '\000', translating=0 '\000', pt=0x555515592478, local_info=0x7ffdf7b1e870, pair_data=0x5555156152b0, quintet_data=0x5555156152f0)
    at /home/bruening/dr/git/src/ext/drmgr/drmgr.c:939
#4  0x000055555590d05e in drmgr_bb_event (drcontext=0x7ffdf7afd080, tag=0x5555558f0051 <replace_retaddr_sentinel>, bb=0x7ffdf7b23138, 
    for_trace=0 '\000', translating=0 '\000') at /home/bruening/dr/git/src/ext/drmgr/drmgr.c:1195
```

@johnfxgalea -- any thoughts on getting drbbdup to handle this no-app-instr block?  drwrap inserts a clean call and then a redirect through DR's IBL in the block, and we need that to only happen for one of the drbbdup cases, so we need drbbdup to duplicate that meta nop as though it were an app instr.  drbbdup_duplicate_phase has this assert which will fail (though it crashes on the line above it with the callstack above:
```C
    app_pc pc = instr_get_app_pc(instrlist_first_app(bb));
    ASSERT(pc != NULL, "pc cannot be NULL");
```

**Comment by derekbruening** [comment]
We have another problem: for the #3995 case with alternating phases of wrap and don't-wrap, we could have a transition to don't-wrap in the middle of a wrapped callee.  With nested wrap targets things could get messed up depending on where the stack pointer is on resumption of wrapping; and with DRWRAP_REPLACE_RETADDR we would hit fatal issues on reaching the sentinel retaddr.  Some kind of drwrap_unwind() that undoes all the levels and restores the retaddr is needed.  And docs about a user's own state, if any, needing to also be reset.

**Comment by johnfxgalea** [comment]
> @johnfxgalea -- any thoughts on getting drbbdup to handle this no-app-instr block? drwrap inserts a clean call and then a redirect through DR's IBL in the block, and we need that to only happen for one of the drbbdup cases, so we need drbbdup to duplicate that meta nop as though it were an app instr. drbbdup_duplicate_phase has this assert which will fail (though it crashes on the line above it with the callstack above:

I'm still thinking....

So a little background information related to the pc. Drbbdup makes that assumption because it uses the pc of the first app instruction as an identifier (key) for the hashtable in order to fetch the corresponding manager for that basic block. Currently, without a pc, we cannot associate a manager to the bb; hence the assumption.

**Comment by derekbruening** [comment]
For the first problem of the dr_emit_flags_t: I am making a PR now that adds a new callback that returns flags (and takes the two bools while at it, just in case).

**Comment by johnfxgalea** [comment]
> For the first problem of the dr_emit_flags_t: I am making a PR now that adds a new callback that returns flags (and takes the two bools while at it, just in case).

On the top of my head, I think that should be fine.

Perhaps, since drbbdup is quite a relatively recent extension, its API could still be considered experimental in relation to whether we should maintain backwards compatibility (of course with proper mention in the release docs). Your call, but worth considering if the change gets messy.

**Comment by derekbruening** [comment]
> > @johnfxgalea -- any thoughts on getting drbbdup to handle this no-app-instr block? drwrap inserts a clean call and then a redirect through DR's IBL in the block, and we need that to only happen for one of the drbbdup cases, so we need drbbdup to duplicate that meta nop as though it were an app instr. drbbdup_duplicate_phase has this assert which will fail (though it crashes on the line above it with the callstack above:
> 
> I'm still thinking....
> 
> So a little background information related to the pc. Drbbdup makes that assumption because it uses the pc of the first app instruction as an identifier (key) for the hashtable in order to fetch the corresponding manager for that basic block. Currently, without a pc, we cannot associate a manager to the bb; hence the assumption.

Would the tag work instead?  Does it need to be a real app PC?

**Comment by johnfxgalea** [comment]
When a basic block is flushed (e.g., due to dynamic handling), we still need to maintain the link to its corresponding manager if it gets created again. AFAIR, tags were unreliable in such cases.

**Comment by derekbruening** [comment]
> When a basic block is flushed (e.g., due to dynamic handling), we still need to maintain the link to its corresponding manager if it gets created again. AFAIR, tags were unreliable in such cases.

Hmm, I don't remember a problem specifically with tags.  The maintenance problem I remember is with dr_register_delete_event as commented by that function: it was only if the client tears down data too early.

If the drbbdup manager data is kept forever: how does it handle modified code?  Or the assumption is that whatever code is at that address, it should have the same drbbdup behavior?

**Comment by johnfxgalea** [comment]
> > When a basic block is flushed (e.g., due to dynamic handling), we still need to maintain the link to its corresponding manager if it gets created again. AFAIR, tags were unreliable in such cases.
> 
> Hmm, I don't remember a problem specifically with tags. The maintenance problem I remember is with dr_register_delete_event as commented by that function: it was only if the client tears down data too early.

I believe it was an issue I faced privately. I tried using tags initially when implementing the first version of drbbdup for my PhD. I don't think I ever documented it publicly as the merged version uses PCs as the key. 

> If the drbbdup manager data is kept forever: how does it handle modified code? Or the assumption is that whatever code is at that address, it should have the same drbbdup behavior?

The manager maintains an `is_gen` flag, which is set to true when performing dynamic handling for its basic block. If we are not creating the basic block again due to the dynamic trigger, we create a fresh manager. Modified code falls under this case.

https://github.com/DynamoRIO/dynamorio/blob/d16f93ef6c8702dfa7709f9ee192bcb4058ce8aa/ext/drbbdup/drbbdup.c#L438

**Comment by johnfxgalea** [comment]
One half-baked solution that I *think* I have is to have another  hashtable for maintaining managers that correspond to basic blocks which do not contain a valid app pc. These managers do not persist and dynamic case generation is not allowed.

I'll think more about it...

**Comment by johnfxgalea** [comment]
One might get around using a hashtable all together for these types of managers and pass them via parameters or thread-local storage (since they don't persist).

**Comment by johnfxgalea** [comment]
@derekbruening The above is no longer an issue? How was it resolved?

**Comment by derekbruening** [comment]
> Another solution might be to leave drwrap post-call actions as always on in every drbbdup case in terms of updating drwrap state, yet not invoking user post-call handlers? Then we always clean up things like changed retaddrs and we don't have to worry about getting the timing right or synchronizing with all other threads.

But how do we skip handlers in all other threads?
Force the bbdup user to check the case encoding in their own post-call handlers?

**Comment by derekbruening** [comment]
> something like #5319 might work if it set a flag in each thread that was read by the drwrap pre code so that the next time that thread does any drwrap operation it will first do the clear.

That's not good enough: for DRWRAP_REPLACE_RETADDR we'll hit the sentinel retaddr and if we don't run the drwrap handling code for that case we'll probably just crash.  It seems a lazy solution is not possible unless we have the drwrap post-call handler in every case.

**Comment by derekbruening** [comment]
We could insert two types of post-call drwrap instru: one normal one that
invokes post-call handlers into the desired drbbdup cases, and one that
just cleans up (and handles the sentinel) into all other drbbdup cases.
Then we don't have to do anything special in drwrap on a phase transition?

**Comment by derekbruening** [comment]
> > > > When a basic block is flushed (e.g., due to dynamic handling), we still need to maintain the link to its corresponding manager if it gets created again. AFAIR, tags were unreliable in such cases.
> > > 
> > > 
> > > Hmm, I don't remember a problem specifically with tags. The maintenance problem I remember is with dr_register_delete_event as commented by that function: it was only if the client tears down data too early.
> > 
> > 
> > I believe it was an issue I faced privately. I tried using tags initially when implementing the first version of drbbdup for my PhD. I don't think I ever documented it publicly as the merged version uses PCs as the key.
> 
> I'm curious what the problem with tags was: were you able to find it in your notes or anything?

I'm using tags locally as I'm blocked on this.  Will do some testing on larger apps next week; though they don't have JITs so maybe they won't hit whatever issue you hit.

**Comment by johnfxgalea** [comment]
The reason is related to this https://groups.google.com/g/dynamorio-users/c/19oX1qOohB8/m/NU4nB2SmAAAJ

**Comment by derekbruening** [comment]
> The reason is related to this https://groups.google.com/g/dynamorio-users/c/19oX1qOohB8/m/NU4nB2SmAAAJ

I guess I'm still not understanding: from DR's point of view, the tag is the unique identifier for a block.  Whether an app2app transformation tweaked the first "app pc" in the block or whether the tag is really a trampoline for a Windows syscall that DR hooks, the tag is guaranteed to be a unique method of referring to just one block (though thread-private blocks can duplicate the same tag per thread).

That sounds like exactly what drbbdup wants for the key to its manager hashtable?  The tag?

**Comment by johnfxgalea** [comment]
AFAIK, the tag is not guaranteed to persist across different bb constructions of the same code; a key requirement for handling dynamic handling as  the manager still needs to be associated with that code when its bb is flushed and reconstructed again.

**Comment by derekbruening** [comment]
It will be the same tag every time for the same source application code.  The tag is independent of what clients do to the instruction list once core DR decodes the source app code (e.g., no matter what app2app transformations are done, the tag won't change).  But the tag will always be the same for the same starting source application address.

**Comment by johnfxgalea** [comment]
Makes sense... Okay, I changed the code back to using tags as a key for the manager hash-table in the new PR (https://github.com/DynamoRIO/dynamorio/pull/5381).

**Comment by derekbruening** [comment]
For a user who has state between pre and post and so needs post called in all phases, I'm thiinking the user would simply not use  this inverted mode.  Then drwrap would use the drmgr events and it should put its calls before the drbbdup dispatch and I think it will all work out.

**Comment by derekbruening** [comment]
I believe everything here is now fixed.  Closing.

**Comment by johnfxgalea** [comment]
> It will be the same tag every time for the same source application code. The tag is independent of what clients do to the instruction list once core DR decodes the source app code (e.g., no matter what app2app transformations are done, the tag won't change). But the tag will always be the same for the same starting source application address.


So tags passed to instrumentation routines for basic blocks and traces are the same? If the same source application code is considered for basic block and trace construction, instrumentation routines in both cases would have the same tag, right? The only difference in terms of parameter values would be the `for_trace` flag, which would be true when constructed the trace?

**Comment by johnfxgalea** [comment]
I see. And sequential basic blocks that make up a trace do not have the same tags as when they are constructed individually for the bb cache?

**Comment by derekbruening** [comment]
> I see. And sequential basic blocks (i.e., not the head) that make up a trace do not have the same tags as when they are constructed individually for the bb cache?

They do have the same tag as for the bb cache.  Was focusing on the head, but they're all the same as the corresponding bb fragment.  And the tags are always app code except for the case of DR hooking raw code when the tag is DR's trampoline code, which DR hides via dr_fragment_app_pc pointing at the clobbered app code.
